erc20/gno/gno-erc20-spec.ini

```diff
@@ -0,0 +1,252 @@
+
+[allowance]
+callData: #abiCallData("allowance", #address(OWNER), #address(SPENDER))
+log: _
+refund: _
+
+[allowance-success]
+statusCode: _ => EVMC_SUCCESS
+output: _ => #asByteStackInWidth(ALLOWANCE, 32)
+callValue: 0
+storage:
+    #hashedLocation({COMPILER}, {_ALLOWED}, OWNER SPENDER) |-> ALLOWANCE
+    _:Map
+requires:
+    andBool 0 <=Int OWNER     andBool OWNER     <Int (2 ^Int 160)
+    andBool 0 <=Int SPENDER   andBool SPENDER   <Int (2 ^Int 160)
+    andBool 0 <=Int ALLOWANCE andBool ALLOWANCE <Int (2 ^Int 256)
+
+[allowance-failure]
+statusCode: _ => EVMC_INVALID_INSTRUCTION
+output: _ => _
+callValue: CALL_VALUE
+storage: _:Map
+requires:
+    andBool 0 <Int  CALL_VALUE andBool CALL_VALUE <Int (2 ^Int 256)
+
+[approve]
+callData: #abiCallData("approve", #address(SPENDER), #uint256(VALUE))
+refund: _ => _
+
+[approve-success]
+statusCode: _ => EVMC_SUCCESS
+output: _ => #asByteStackInWidth(1, 32)
+log: _:List ( .List => ListItem(#abiEventLog(ACCT_ID, "Approval", #indexed(#address(CALLER_ID)), #indexed(#address(SPENDER)), #uint256(VALUE))) )
+callValue: 0
+storage:
+    #hashedLocation({COMPILER}, {_ALLOWED}, CALLER_ID SPENDER) |-> (_:Int => VALUE)
+    _:Map
+requires:
+    andBool 0 <=Int SPENDER   andBool SPENDER   <Int (2 ^Int 160)
+    andBool 0 <=Int VALUE     andBool VALUE     <Int (2 ^Int 256)
+
+[approve-failure]
+statusCode: _ => EVMC_INVALID_INSTRUCTION
+output: _ => _
+log: _
+callValue: CALL_VALUE
+storage: _:Map
+requires:
+    andBool 0 <Int  CALL_VALUE andBool CALL_VALUE <Int (2 ^Int 256)
+
+[transfer]
+callData: #abiCallData("transfer", #address(TO_ID), #uint256(VALUE))
+refund: _ => _
+requires:
+    andBool 0 <=Int TO_ID     andBool TO_ID     <Int (2 ^Int 160)
+    andBool 0 <=Int VALUE     andBool VALUE     <Int (2 ^Int 256)
+    andBool 0 <=Int BAL_FROM  andBool BAL_FROM  <Int (2 ^Int 256)
+    andBool 0 <=Int BAL_TO    andBool BAL_TO    <Int (2 ^Int 256)
+
+[transfer-success]
+statusCode: _ => EVMC_SUCCESS
+output: _ => #asByteStackInWidth(1, 32)
+log: _:List ( .List => ListItem(#abiEventLog(ACCT_ID, "Transfer", #indexed(#address(CALLER_ID)), #indexed(#address(TO_ID)), #uint256(VALUE))) )
+callValue: 0
+
+[transfer-success-1]
+storage:
+    #hashedLocation({COMPILER}, {_BALANCES}, CALLER_ID) |-> (BAL_FROM => BAL_FROM -Int VALUE)
+    #hashedLocation({COMPILER}, {_BALANCES}, TO_ID)     |-> (BAL_TO   => BAL_TO   +Int VALUE)
+    _:Map
++requires:
+    andBool CALLER_ID =/=Int TO_ID
+    andBool VALUE <=Int BAL_FROM
+    andBool BAL_TO +Int VALUE <Int (2 ^Int 256)
+
+[transfer-success-2]
+storage:
+    #hashedLocation({COMPILER}, {_BALANCES}, CALLER_ID) |-> BAL_FROM
+    _:Map
++requires:
+    andBool CALLER_ID ==Int TO_ID
+    andBool VALUE <=Int BAL_FROM
+
+[transfer-failure]
+output: _ => _
+log: _
+
+[transfer-failure-1]
+statusCode: _ => EVMC_REVERT
+callValue: 0
+storage:
+    #hashedLocation({COMPILER}, {_BALANCES}, CALLER_ID) |-> (BAL_FROM => _)
```

Comment:
BAL_FROM should not change in case of failure. Either incomplete specification or contract not compliant with ERC20. Same for other failure specs.

Reply:
Yeah I raised this before (https://gnosisinc.slack.com/archives/CCDAA679Q/p1539687195000100 -- let me know if you can view that).

I agree that it's counter-intuitive, but it's correct (part of the ethereum protocol is reverting all changes if a call leads end with REVERT. I copied the above from the Vyper erc20 spec. Let me know how to proceed 👍

Reply:
Then why you need a rewrite in storage if all changes are reverted? I can't see the link btw.